### DISCUSSION TITLE
How to get the Type of a monitored item?

### DISCUSSION THREAD

**Discussion by ErikTempelaarVO** [open62541pp/open62541pp#273] [Q&A] [answered]
I have a client that subscribes to a set of monitored items without knowing their type in advance.

Before I was able to get the type from the DataValue via its variant.
Something like below:
```
void SomeClass::WriteTag(const opcua::ReadValueId& rvid, const ::opcua::DataValue& value) noexcept {
  const auto& variant = value.getValue();
  auto optType = variant.getVariantType();
  switch (optType.value()) {
    case opcua::Type::Boolean:
      // get the value
      variant.getScalarCopy<bool>();
      break;
    case opcua::Type::Byte:
      ...
      break;
    case opcua::Type::SByte:
      ...
      break;

```

This approach is now deprecated. Type-enum is on its way out and should be replaced by DataTypeId?
```
    /// Get variant type.
    /// @deprecated Use getDataType() or isType<T>() instead, the Type enum will be removed
    [[deprecated("Use getDataType() or isType<T>() instead, the Type enum will be removed")]]
    std::optional<Type> getVariantType() const noexcept {
        if (handle()->type != nullptr) {
            const auto typeIndex = handle()->type->typeKind;
            if (typeIndex <= UA_DATATYPEKIND_DIAGNOSTICINFO) {
                return static_cast<Type>(typeIndex);
            }
        }
        return {};
    }
```

I think it should use getDataType(), because calling isType<T> for several types sounds tedious.
But this returns a UA_DataType* and I'm unsure how to proceed from there? Any advice?

**Comment by lukasberbuer** [comment]
Hi @ErikTempelaarVO,

I would try to understand your use-case a little bit more in detail.
Do you have a list of corresponding types you would like to read? E.g. `bool`, `int`, ...?
Do you perform different actions depending on the type?

Initially, the `Types` enum was meant to describe possible types of the `Variant` class, as defined in the OPC UA standard. But open62541 has a relaxed concept of variants and allows users to store any known data type in a variant. Therefore, I didn't see any advantage of keeping the `Types` enum...

**Comment by ErikTempelaarVO** [comment]
Thanks for your reply @lukasberbuer 

Yes, you are correct, the client has a limited set of types it needs to support and has to get their values and store them elsewhere, with type information in a different format.
It should be able to deal with these monitored items, regardless of their type.

Currently I have a type-mapping in place:
```constexpr CustomNameType getCustomNameType(opcua::Type type) {
  using ::opcua::Type;
  switch (type) {
    case Type::Boolean:
      return CustomNameType::Boolean;
    case Type::Byte:
      return CustomNameType::U_INT_8;
    case Type::SByte:
      return CustomNameType::S_INT_8;
    case Type::Int16:
      return CustomNameType::S_INT_16;
    case Type::UInt16:
      return CustomNameType::U_INT_16;
    case Type::Int32:
      return CustomNameType::S_INT_32;
    case Type::UInt32:
      return CustomNameType::U_INT_32;

    ... a handful more here ...

  }
}
```
Then `getScalarCopy<T>` is used to get the value, depending on the type, like below.

This was already a bit clunky, and I guess there's a better way to do this, via TypeConverter or TypeRegistry, but I haven't figured that part out yet.

Ofcourse I could use UA_DataType in the switch, but I'd rather interface with open62541pp.

```
    switch (optType.value()) {
      case opcua::Type::Boolean:
        WriteTag(tagName, Transform_t(variant.getScalarCopy<bool>()));
        break;
      case opcua::Type::Byte:
        WriteTag(tagName, Transform_t(variant.getScalarCopy<uint8_t>()));
        break;
      case opcua::Type::SByte:
        WriteTag(tagName, Transform_t(variant.getScalarCopy<int8_t>()));
        break;
      case opcua::Type::Int16:
        WriteTag(tagName, Transform_t(variant.getScalarCopy<int16_t>()));
        break;
      case opcua::Type::UInt16:
        WriteTag(tagName, Transform_t(variant.getScalarCopy<uint16_t>()));
        break;
      case opcua::Type::Int32:
        WriteTag(tagName, Transform_t(variant.getScalarCopy<int32_t>()));
        break;
 ```
 
 Hope this helps you understand what I'm trying to achieve. Please let me know if there's better way to do this.

**Comment by lukasberbuer** [comment, marked as answer]
Shouldn't it be better to use the template type for dispatching instead of enums and long switch statements?
I was thinking about a visit design pattern as known from `std::variant`. But open62541's variant type can hold any type if a `UA_DataType` description is provided. So we need to restrict the open set of types to a closed set.

A possible visit implementation could look like this:

```cpp
#include <functional>  // invoke
#include <iostream>
#include <utility>  // forward

#include "open62541pp/open62541pp.h"

template <typename... Ts>
struct Overload : Ts... { using Ts::operator()...; };
template <typename... Ts>
Overload(Ts...) -> Overload<Ts...>;

template <typename T, typename F>
constexpr bool visitScalarHelper(F&& visitor, const opcua::Variant& var) {
    if (var.isType<T>() && var.isScalar()) {
        std::invoke(std::forward<F>(visitor), var.getScalar<T>());
        return true;
    }
    return false;
}

template <typename... Ts, typename F>
constexpr bool visitScalar(F&& visitor, const opcua::Variant& var) {
    return (visitScalarHelper<Ts>(std::forward<F>(visitor), var) || ...);
}

int main() {
    auto var = opcua::Variant::fromScalar(11.11);
    const bool visited = visitScalar<bool, float, double>(
        Overload{
            [](bool value) { std::cout << "bool: " << value << std::endl; },
            [](float value) { std::cout << "float: " << value << std::endl; },
            [](double value) { std::cout << "double: " << value << std::endl; },
            [](auto&& value) { std::cout << "anything: " << value << std::endl; },
        },
        var
    );
    std::cout << "visited: " << visited << std::endl;
}
```

What do you think? Could this fit your use-case?

**Comment by ErikTempelaarVO** [comment]
Thanks, that visitor is definitely helpful.

**Comment by lukasberbuer** [comment]
Please let me know if you see any advantage of the enum-based dispatch... we might reconsider deprecating the `Type` enum.
